### ISSUE TITLE
SSL handshake failure when connecting many concurrent connections (C# server)

### ISSUE THREAD

**Issue by amitwaisel** [grpc/grpc#8382]
@apolcyn
Using a C# server listening to receive **SSL** conections from C++/C# clients, using `SslServerCredentials` (root certificate, certificate chain and private key on both sides), Most of the connections fail to be initialized (or get disconnected after a few seconds, in an early stage, while printing the following message to stdout:

> E1016 12:43:43.415940 0 c:\jenkins\workspace\gRPC_build_artifacts\architecture\x86\language\csharp\platform\windows\vsprojects..\src\core\lib\security\transport\handshake.c:128: Security handshake failed: {"created":"@1476611023.414000000","description":"Handshake read failed","file":"c:\jenkins\workspace\gRPC_build_artifacts\architecture\x86\language\csharp\platform\windows\vsprojects..\src\core\lib\security\transport\handshake.c","file_line":237,"referenced_errors":[{"created":"@1476611023.414000000","description":"End of TCP stream","file":"c:\jenkins\workspace\gRPC_build_artifacts\architecture\x86\language\csharp\platform\windows\vsprojects..\src\core\lib\iomgr\tcp_windows.c","file_line":180}]}

Both clients and server are running from a Windows machine (particularly Win10 x64). The C# code uses GRPC v1.0.0 from NuGet, C++ clients use the latest revision from v1.0.x branch.

When using an insecure channel, this error doesn't exist and connections are established successfully.
Can it be related to `ChannelOptions.MaxConcurrentStreams`? Although the secure and insecure channels have the same default value, and only the secure channels get disconnected.

**Comment by apolcyn** [comment]
Making multiple calls on concurrent threads, from a C# client to C# server, I think I am seeing the same error.

Seeing the error on the client side from the client calls, and relevant stack trace of:
```
    E1018 14:03:39.825119 0 c:\jenkins\workspace\gRPC_build_artifacts\architecture\x86\language\csharp\platform\windows\
vsprojects\..\src\core\lib\security\transport\handshake.c:128: Security handshake failed: {"created":"@1476824619.824000
000","description":"

Handshake read failed","file":"c:\jenkins\workspace\gRPC_build_artifacts\architecture\x86\language\csharp\platform\windo
ws\vsprojects\..\src\core\lib\security\transport\handshake.c","file_line":237,"referenced_errors":[{"created":"@14768246
19.824000000","description":"End of TCP stream","file":"c:\jenkins\workspace\gRPC_build_artifacts\architecture\x86\langu
age\csharp\platform\windows\vsprojects\..\src\core\lib\iomgr\tcp_windows.c","file_line":180}]}
Unhandled Exception:Grpc.Core.RpcException: Status(StatusCode=Unavailable, Detail="")
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Grpc.Core.Internal.AsyncCall`2.UnaryCall(TRequest msg)
   at Grpc.Core.Calls.BlockingUnaryCall[TRequest,TResponse](CallInvocationDetails`2 call, TRequest req)
   at Grpc.Core.DefaultCallInvoker.BlockingUnaryCall[TRequest,TResponse](Method`2 method, String host, CallOptions optio
ns, TRequest request)
   at Grpc.Core.Internal.InterceptingCallInvoker.BlockingUnaryCall[TRequest,TResponse](Method`2 method, String host, Cal
lOptions options, TRequest request)
   at Helloworld.Greeter.GreeterClient.SayHello(HelloRequest request, CallOptions options) in C:\Users\apolcyn\grpc_work
space\grpc\examples\csharp\helloworld\Greeter\HelloworldGrpc.cs:line 112
   at Helloworld.Greeter.GreeterClient.SayHello(HelloRequest request, Metadata headers, Nullable`1 deadline, Cancellatio
nToken cancellationToken) in C:\Users\apolcyn\grpc_workspace\grpc\examples\csharp\helloworld\Greeter\HelloworldGrpc.cs:l
ine 105
   at GreeterClient.Program.<>c__DisplayClass0_0.<Main>b__0() in C:\Users\apolcyn\grpc_workspace\grpc\examples\csharp\he
lloworld\GreeterClient\Program.cs:line 61
   at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
```

this seems to be triggered by making enough new secure connections concurrently, with something like:
```
            for (int i = 0; i < 300; i++)
            {
                Channel newChan = new Channel("127.0.0.1:50051", TestCredentials.CreateSslCredentials(), options);
                channels.Add(newChan);
                var k = i;

                var thread = new System.Threading.Thread(new System.Threading.ThreadStart(() =>
                {
                    var client = new Greeter.GreeterClient(newChan);
                    var response = client.SayHello(new HelloRequest { Name = "user " + k.ToString() });
                    Console.WriteLine("Hello: " + response.Message);
                }));
                threads.Add(thread);
                thread.Start();
            }
            threads.ForEach((t) => t.Join());
```

on the client, and a secure, Greeter server from the example.

With 300 such threads, I almost always get this error. Fewer threads than that and it will sometimes happen, but more rarely. 

With insecure connections, many more channels can be used without issues (had no issues with 1000, but I start seeing Unavailable errors if it's stepped up too high, but I think this is expected).

**Comment by apolcyn** [comment]
the same error, with the same observations as above, also happens on OS X with mono 4.4.2

**Comment by apolcyn** [comment]
@amitwaisel  so this looks like it's a current problem with the retry logic around SSL handshakes.

For your current situation, try setting the "grpc.initial_reconnect_backoff_ms" argument when constructing the client `Channel`.

For example, this fixed the "handeshake failure exceptions" problem with the repro attempts shown above in this issue:

```
var options = new List<ChannelOption>
{
    new ChannelOption(ChannelOptions.SslTargetNameOverride,
        TestCredentials.DefaultHostOverride),
    new ChannelOption("grpc.initial_reconnect_backoff_ms", "100"),
};

Channel channel = new Channel("127.0.0.1:50051", TestCredentials.CreateSslCredentials(), options);
```

Setting the backoff too a longer period might be necessary, probably depending on the load against the server.

It's default value is actually zero as of right now, though this may change - I think raising it at all should fix the immediate issue.

**Comment by amitwaisel** [comment]
@apolcyn This bug (or similar) still exists in the released 1.0.1 version.

> E1107 16:02:06.265602 0 c:\jenkins\workspace\gRPC_build_artifacts\architecture\x86\language\csharp\platform\windows\vsprojects\..\src\core\lib\security\transport\handshake.c:128: Security handshake failed: {"created":"@1478527326.265000000","description":"Handshake read failed","file":"c:\jenkins\workspace\gRPC_build_artifacts\architecture\x86\language\csharp\platform\windows\vsprojects\..\src\core\lib\security\transport\handshake.c","file_line":237,"referenced_errors":[{"created":"@1478527326.265000000","description":"**An established connection was aborted by the software in your host machine**.\r\n","file":"c:\jenkins\workspace\gRPC_build_artifacts\architecture\x86\language\csharp\platform\windows\vsprojects\..\src\core\lib\iomgr\tcp_windows.c","file_line":171}]}
E1107 16:02:06.268611 0 c:\jenkins\workspace\gRPC_build_artifacts\architecture\x86\language\csharp\platform\windows\vsprojects\..\src\core\ext\transport\chttp2\server\secure\server_secure_chttp2.c:119: Secure transport failed with error 1

After getting this error, all further connections fail with this error:
> E1107 16:02:06.271631 0 c:\jenkins\workspace\gRPC_build_artifacts\architecture\x86\language\csharp\platform\windows\vsprojects\..\src\core\lib\security\transport\handshake.c:128: Security handshake failed: {"created":"@1478527326.271000000","description":"Handshake read failed","file":"c:\jenkins\workspace\gRPC_build_artifacts\architecture\x86\language\csharp\platform\windows\vsprojects\..\src\core\lib\security\transport\handshake.c","file_line":237,"referenced_errors":[{"created":"@1478527326.271000000","description":"**End of TCP stream**","file":"c:\jenkins\workspace\gRPC_build_artifacts\architecture\x86\language\csharp\platform\windows\vsprojects\..\src\core\lib\iomgr\tcp_windows.c","file_line":180}]}
E1107 16:02:06.273623 0 c:\jenkins\workspace\gRPC_build_artifacts\architecture\x86\language\csharp\platform\windows\vsprojects\..\src\core\ext\transport\chttp2\server\secure\server_secure_chttp2.c:119: Secure transport failed with error 1

This issue **also happens after adding the ChannelOption** `grpc.initial_reconnect_backoff_ms` as you instructed.

**Comment by jtattermusch** [comment]
Based on my research, this seems to be caused (at least partially) by the connection attempts timing out.

When a subchannel is created, it is given a gpr_backoff object that  determines the connection attempt deadline and the time for the next attempt (both are actually the same value, so the first attempt to connect is given a too short limit).  If SSL handshake doesn't finish on time, the connection is closed (by on_timeout event in handshake.c). This seems to explain the insecure/secure difference, because the SSL handshake takes several roundtrips and is much more likely to miss the deadline, especially when there are more connections created in parallel.

For the time being, the workaround is to set grpc.initial_reconnect_backoff_ms to a HIGHER value (the default is 1000). I tried setting it to 10000 (so the first connection attempt for each channel will have a timeout of 10s) and I stopped seeing most of the failures to connect (I can connect 100 channels in parallel without a problem).

IMPORTANT: the channel argument `grpc.initial_reconnect_backoff_ms` is of integer type and it's value needs to be an integer.

```
// WRONG:  sets string value to an integer argument
// VALUE WILL BE IGNORED!!!
new ChannelOption("grpc.initial_reconnect_backoff_ms", "10000")

// CORRECT
new ChannelOption("grpc.initial_reconnect_backoff_ms", 10000)
```
(I will file a separate issue for the channel arg being ignored, as the behavior is very confusing)

**Comment by jtattermusch** [comment]
An actual fix for this will likely be resolving #8696

**Comment by amitwaisel** [comment]
@jtattermusch Your solution appears to solve the majority of the errors (as I understand, this channel argument is relevant to the client side)
Under heavy load, encrypted connections start to fail after a while (over a minute) with the following `RpcException`:
> Status(StatusCode=Cancelled, Detail="{"created":"@1478759465.033000000","description":"RST_STREAM","file":"c:\jenkins\workspace\gRPC_build_artifacts\architecture\x86\language\csharp\platform\windows\vsprojects\..\src\core\ext\transport\chttp2\transport\frame_rst_stream.c","file_line":107,"http2_error":8}")

Which means some connection got RST packets.
I found that many channel arguments are declared in `grpc_types.h`, with a little comment above each one of them, describing its purpose. Do you have some more detailed documentation for those arguments? It might solve other problems (like the RST packets)

**Comment by jtattermusch** [comment]
@amitwaisel   let's leave this issue open until #8696  because currently our defaults are broken and one needs to set initial_reconnect_backoff_ms to work around it.

Also, would you mind filing the RST_STREAM problem as a separate issue? It looks like a different problem and should be investigated separately. Thanks!
